### ISSUE TITLE
bump event-pubsub to v5+ from 4.3.0 and use `.reset()` functionality

### ISSUE THREAD

**Issue by tenbits** [RIAEvangelist/node-ipc#187]
Hi Team,

current version uses `event-pubsub@4.3.0` [package.json#L13](https://github.com/RIAEvangelist/node-ipc/blob/c82afb34b504e1ff9204c01d457ba645aa7935da/package.json#L13), there for `this.of[id].off('*','*');` in [IPC.js#L95](https://github.com/RIAEvangelist/node-ipc/blob/c82afb34b504e1ff9204c01d457ba645aa7935da/services/IPC.js#L95) does nothing, as such "unsubscribe all" was not implemented in [event-pubsub@4.3.0/es6.js](https://unpkg.com/event-pubsub@4.3.0/es6.js) as the glob `*` is not supported for a `type`.  
In newer version this is already present [event-pubsub@5.0.4/index.js#L40](https://github.com/RIAEvangelist/event-pubsub/blob/e45fb89ba0965259fe949b0a1e693cdb34a3e4b6/index.js#L40)

**Comment by RIAEvangelist** [comment]
Interesting, I will have a look.

**Comment by RIAEvangelist** [comment]
can you provide a code sample to confirm this? I wonder if you are confusing the `*` event which is triggered on any event with the glob `*` which means all events. These are very different.

```js
const allHandler=(){console.log('all')};
const specificHandler=(){console.log('specific')};

// * event handlers are triggered on ALL events
x.on('*',allHandler)
//named events are only triggered for their specific event
x.on('specificEvent',specificHandler)

x.emit('specificEvent')

//this will log "all" & "specific"

x.emit('randomEvent')

//this will log just "all" from the  * event

x.off('*','*')

//this will remove all '*' events and handlers

x.emit('specificEvent')
//this will now log just "specific" because the '*' events are all gone


```

I see in `event-pubsub` v4.3.0/es6 that you linked, the following code which looks to behave exactly as described above.

```js

off( type, handler ) {
        
         ...

        if ( handler == '*' ) {
            delete this._events_[ type ];
            return this;
        }

        ...
        
    }
    
    ```
    
    Is it possible you just misunderstood the goal of the line you mentioned in the IPC class?

**Comment by RIAEvangelist** [comment]
I checked the v5.0 code refrence, and it looks like it does the same thing to me.  If you can provide a code sample showing what you are expecting I would consider further.

Thank you for your help!

**Comment by RIAEvangelist** [comment]
I took a little deeper look into what I think you are expecting which 5.0 `event-pubsub` does have as the `.reset()` method, which is not related to the line you linked as the `.#all` event type is a special event type to be ***triggered*** for all events. 

I do not think that line in the IPC code is doing what you expect.

**Comment by RIAEvangelist** [comment]
I reviewed the IPC code as well, and I am not sure that the line you referenced in IPC is even valuable because the remainder of the code just removes the whole thing:

```js
    
    // seems somewhat pointless
    this.of[id].off('*','*');
    
    if(this.of[id].socket){
        if(this.of[id].socket.destroy){
            //destroy here
            this.of[id].socket.destroy();
        }
    }

    //delete here
    delete this.of[id];

```

Is there any other reason to update to the new version of `event-pubsub`?

**Comment by tenbits** [comment]
Hi Brandon, thanks a lot in getting into this, yes I meant that 

```js
// seems somewhat pointless
this.of[id].off('*','*');

/**
 * This supposed to remove all events and handlers right? 
 * But in event-pubsub@4.3.0/es6.js 
*/

off( type, handler ) {
    if ( !this._events_[ type ] ) {
        return this;
    }
    //..
}

/** So '*' not supported as a glob for the `type` - type should be specific event name
```

Actually I stumbled across this in another quit tricky case:

1. Connect as a client to unix/win socket server (_in another process_)
2. In `disconnect` event (_when another process exits_), close the client with `ipc.disconnect(id)`
3. Create new unix/win socket server with `ipc.serve`, so that the process acts now not as a client, but as a host.

And the issue was, I haven't disabled the retries, that's why even though I was calling `ipc.disconnect(id)` in `disconnect` event, the retry was already in progress, and after new server was immediately created in step 3. the socket successfully connects and the events were receiving, and so here I noticed that the event unsubscription was not functional, but this helped me to notice the retry issue with the client socket.

Sorry Brandon, my explanation could be not quit well. But anyway, I could fix my case by disabling the retries for the client. 

And thank you for your time. You are free to close the issue, as the update of the `event-pubsub` is far away from being critical.

**Comment by RIAEvangelist** [comment]
Before I close the issue I want to make sure that the issue you had was legitimate and not a different problem that should be addressed. Also, I am not against updating the `event-pubsub` I just want to make sure I understand the reason, even if that is just because it is newer. I just like to understand first.

Are you sure you had the `id` when you called `ipc.disconnect(id)` and that it matched `ipc.of[id]` in the `disconnect` event? I know there is a current issue where the id is not actually passed to the disconnect event sadly. This is why I made #188 

I would just like to confirm that you do infact have the id where you are calling it, and it is the right id for `ipc.of`.

Also, you are welcome! Thank you for giving good feedback! 🎈

**Comment by tenbits** [comment]
Right, the `id` is correct and the `ipc.of[id]` is present. `ipc.disconnect(id)` indeed destroys the socket and deletes `ipc.of[id]`. But let me go through the steps:

1. When remote host exits and the connection is closed, but retries are **active** the code goes to [client.publish('disconnect')](https://github.com/RIAEvangelist/node-ipc/blob/c82afb34b504e1ff9204c01d457ba645aa7935da/dao/client.js#L197)
2. In `disconnect` I'm calling `ipc.disconnect(id)` which destroys socket, also I'm creating new server with `ipc.serve`. (_Now the process is acting not as a client, but as a host_)
3. But **previously** [retryTimeout](https://github.com/RIAEvangelist/node-ipc/blob/c82afb34b504e1ff9204c01d457ba645aa7935da/dao/client.js#L189) was already scheduled in [client.connect()](https://github.com/RIAEvangelist/node-ipc/blob/c82afb34b504e1ff9204c01d457ba645aa7935da/dao/client.js#L192) and new client is being created again and connects to  newly created server from step 2.

**Comment by RIAEvangelist** [comment]
OK, would it be possible for you to either fork the repo and update it to the newer `event-pubsub`, and edit that line to use the `this.of[id].reset()` which should clear all events for `event-pubsub` v5+

OR, just forcibly install it on your own and hack the `.reset` in and confirm it fixes your issue?

If you can confirm it fixes the issue, that will be enough for me to do a new release with the newer event-pubsub.

**Comment by RIAEvangelist** [comment]
Also, if you do a pr for it, you will get credit for being a contributor if you care.

**Comment by tenbits** [comment]
Hi Brandon, I have created for now another pr, that was more critical, which leads to socket connection leaks. I tried also to update the `event-pubsub` module, but `istanbul` test fails, as `event-pubsub` is a es6 module with imports, and instrumenting for some reason doesn't work with the error:

```
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: D:\tenbits\node-ipc\node_modules\event-pubsub\index.js
require() of ES modules is not supported.
require() of D:\tenbits\node-ipc\node_modules\event-pubsub\index.js from D:\tenbits\node-ipc\dao\client.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from D:\tenbits\node-ipc\node_modules\event-pubsub\package.json.
```

**Comment by RIAEvangelist** [comment]
@tenbits I'll update your pr with [`c8`](https://github.com/bcoe/c8) way better than anything else on the market

**Comment by RIAEvangelist** [comment]
ah crap. I want to replace jasmine with [`vanilla-test`](https://www.npmjs.com/package/vanilla-test) too if we're going to upgrade the `event-pubsub` and use `c8` instead of istanbul.

**Comment by RIAEvangelist** [comment]
This is on the way

**Comment by RIAEvangelist** [comment]
Current on master.

Once testing is complete this will be in v10.x on npm

**Comment by joshuaalayon** [comment]
Be aware - this package was found to contain malware that detects its users machines purported geographical location and replaces the data from the machine if it does not like the location it detects

If you are still using this package - you may do better by locking it to an older version or, preferably, replacing it altogether along with all software maintained by the repos maintainer(s)

https://www.bleepingcomputer.com/news/security/big-sabotage-famous-npm-package-deletes-files-to-protest-ukraine-war/amp/
